### PR TITLE
Add useAsync hook

### LINKED ISSUE #86: useAsync hook
---
title: useAsync
tags: react,intermediate,hook
---

The `useAsync` hook is used to bind asynchronous call and handle different states like loading, error, and value. According to the state of an asynchronous call, we can render different states like loading, error and value state.  

**Explanation:**
In `useAsync` hook we need to pass the handler function and that will be called after calling the `run` method. Once `run` get's called we're changing loading, error and value field according to asynchronous call response.

```js
const useAsync = (fn, options = {}) => {
  const autoRun = options.autoRun || false

  const [state, setState] = React.useState({
    error: null,
    loading: false,
    value: null,
  })

  const handleStateChange = args => {
    setState({ ...state, ...args })
  }

  const run = async (args = null) => {
    try {
      handleStateChange({ loading: true, error: null })
      const value = await fn(args)
      handleStateChange({ loading: false, value: value || null, error: null })
    } catch (error) {
      handleStateChange({ loading: false, value: null, error })
    }
  }

  React.useEffect(() => {
    if (autoRun) {
      run()
    }
  }, [autoRun])

  return {
    state,
    setState,
    run,
  }
}
```

**Usages**
```jsx
const App = () => {
  const handleSubmit = () => {
    const url = "https://jsonplaceholder.typicode.com/todos"
    return fetch(url).then(response => response.json())
  }

  const submission = useAsync(handleSubmit, { autoRun: false })

  return (
    <div>
      {submission.value && <div>{submission.value}</div>}
      <button
        onClick={() => submission.run({ foo: "bar" })}
        disabled={submission.loading}
      >
        click me
      </button>
    </div>
  )
}
```

### PR DESCRIPTION
**Pull request by sagar-gavhane**
Resolves #86 

The `useAsync` hook is used to bind asynchronous call and handle different states like loading, error, and value. According to the state of an asynchronous call, we can render different states like loading, error and value state.

## Description
In `useAsync` hook we need to pass the handler function and that will be called after calling the `run` method. Once `run` get's called we're changing loading, error and value field according to asynchronous call response.

## PR Type
- [x] Snippets, Tests & Tags (new snippets, updated snippets, re-tagging of snippets, added/updated tests)
- [ ] Tools, Scripts & Automation (anything related to files in the scripts folder, Gatsby, website, Travis CI or Netlify)
- [ ] General, Typos, Misc. & Meta (everything related to content, typos, general stuff and meta files in the repository - e.g. the issue template)
- [ ] Other (please specify in the description above)

## Guidelines
- [x] I have read the guidelines in the [CONTRIBUTING](https://github.com/30-seconds/30-seconds-of-react/blob/master/CONTRIBUTING.md) document.

### CODE REVIEW

File: snippets/UseAsync.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
+---
+title: useAsync
+tags: react,intermediate,hook
+---
+
+
+The `useAsync` hook is used to bind asynchronous call and handle different states like loading, error, and value. According to the state of an asynchronous call, we can render different states like loading, error and value state.
+
+**Explanation:**
+In `useAsync` hook we need to pass the handler function and that will be called after calling the `run` method. Once `run` get's called we're changing loading, error and value field according to asynchronous call response.
```
Comment (Chalarangelo) [inline snippets/UseAsync.md]: Please follow the format specified in the snippet template and the contribution guidelines.

File: snippets/UseAsync.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
+---
+title: useAsync
+tags: react,intermediate,hook
+---
+
+
```
Comment (Chalarangelo) [inline snippets/UseAsync.md]: Single empty line, instead of two.

File: snippets/UseAsync.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+      <button
+        onClick={() => submission.run({ foo: "bar" })}
+        disabled={submission.loading}
+      >
+        click me
+      </button>
+    </div>
+  )
+}
+```
```
Comment (Chalarangelo) [inline snippets/UseAsync.md]: Please follow the format as specified in the snippet template and contribution guidelines.

File: snippets/UseAsync.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
+---
+title: useAsync
+tags: react,intermediate,hook
```
Comment (Chalarangelo) [inline snippets/UseAsync.md]: Tags should not contain the language. Instead refactor to:
```suggestion
tags: hooks,state,effect,intermediate
```

**Comment by sagar-gavhane** [comment]
@Chalarangelo I've updated PR according to contribution guidelines. Can you please review the latest PR. 

Thanks 👍🏻

File: snippets/UseAsync.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
+---
+title: useAsync
+tags: hooks,state,effect,intermediate
+---
+
+A hook that handle asynchronous calls.
```
Comment (Chalarangelo) [inline snippets/UseAsync.md]:
```suggestion
A hook that handles asynchronous calls.
```

File: snippets/UseAsync.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  return (
+    <div>
+      <button onClick={() => submission.run({ foo: "bar" })} disabled={submission.isLoading}>
+        {submission.isLoading ? 'Loading...': 'click me'}
+      </button>
+      <pre>{JSON.stringify(submission.value, null, 2)}</pre>
+    </div>
+  )
+}
+```
```
[the remark this answers is not in the document]
Comment (sagar-gavhane) [reply snippets/UseAsync.md]: I've added semicolons.

**Comment by Chalarangelo** [comment]
I made some liberal changes to make this work in a more interesting manner, while also differentiating it from [useFetch](https://www.30secondsofcode.org/react/s/use-fetch) and providing some interesting pointers for people looking to get into `React.useReducer()`. Merging...

**Comment by sagar-gavhane** [comment]
Thanks for merging PR.

**Pull request merged**